backend/data_validation/data_val_2022.py

```diff
@@ -126,3 +133,111 @@ def check_for_missing_shooting_zones(
+        :param final_climb_type: Represents the rung the scouted team got to in endgame, if any.
+        :return: None
+        """
+
+        submission_key = match_key.strip().lower()
+        full_match_key = f"{self._event_key}_{submission_key}"
+        score_info = self.tba_match_data[full_match_key]["score_breakdown"]
+
+        if score_info:
+            tba_taxi = score_info[alliance.lower()][f"taxiRobot{driver_station}"]
+            tba_climb = score_info[alliance.lower()][f"endgameRobot{driver_station}"]
+
+            # check for inconsistent taxi
+            if (tba_taxi == "No" and (notna(taxied) or taxied)) or (
+                tba_taxi == "Yes" and (isna(taxied) or not taxied)
+            ):
+                self.add_error(
+                    f"In {match_key}, {team_number} INCORRECT TAXI according to TBA",
+                    error_type=ErrorType.INCORRECT_DATA,
+                )
+
+            # check for inconsistent climb type
+            if tba_climb != final_climb_type.replace("No Climb", "None"):
+                self.add_error(
+                    f"In {match_key}, {team_number} INCORRECT ClIMB TYPE according to TBA, should be {tba_climb}",
+                    error_type=ErrorType.INCORRECT_DATA,
+                )
+
+    def check_for_invalid_defense_data(
+        self,
+        match_key: str,
+        team_number: int,
+        defense_pct: float,
+        defense_rating: float,
+        counter_defense_pct: float,
+        counter_defense_rating: float,
+    ) -> None:
+        """
+        Checks if scouter gave defense or counter defense rating but stated that the robot didn't play defense/counter defense.
+        Checks if scouter stated that robot played defense or counter defense but didn't give a corresponding rating.
+        Checks if total defense played + counter defense played > 100% hence impossible.
+
+        :param match_key: Key of match that was scouted.
+        :param team_number: Number of team that was scouted (eg 4099).
+        :param defense_pct: Decimal representing how much the scouted team played defense out of 1 (eg 0.75)
+        :param defense_rating: Representing how well the scouted team played defense on a scale of 1 to 5.
+        :param counter_defense_pct: Decimal representing how much the scouted team played counter defense out of 1 (eg 0.75)
+        :param counter_defense_rating: Representing how well the scouted team played counter defense on a scale of 1 to 5.
+        :return: None
+        """  # noqa
+        # Check for missing defense pct.
+        if notna(defense_rating) and isna(defense_pct):
+            self.add_error(
+                f"In {match_key}, {team_number} rated for defense but NO DEFENSE PCT",
+                error_type=ErrorType.MISSING_DATA,
+            )
+
+        # Check for missing defense rating.
+        if isna(defense_rating) and notna(defense_pct):
```

Comment:
need to make sure that `defense_rating` is passed in as NaN (or null in submission). i think it's still worth implementing a check here to make sure `defense_rating` is not 0 as a safety net

Reply:
applies to similar checks below too

Reply:
solved